### PR TITLE
Shreya/megatron bridge swap

### PR DESCRIPTION
**Pull request by ShreyaRavi**

### CODE REVIEW

File: trainers/server/src/trainers_server/dp_worker/api/server.py [hunk 1]
```diff
@@ -48,7 +49,7 @@ async def forward_backward(details: ForwardBackwardDetails) -> ForwardBackwardRe
         return _run_or_raise(controller.forward_backward, details)
 
     @app.post("/optim_step", response_model=OptimStepResult)
-    async def optim_step(details: OptimStepDetails = OptimStepDetails()) -> OptimStepResult:
+    async def optim_step(details: OptimStepDetails = OptimStepDetails(adam_params=AdamParams())) -> OptimStepResult:
```
Comment (rcano-baseten) [inline trainers/server/src/trainers_server/dp_worker/api/server.py:52]: does the Tinker API require AdamParams? If so, let's not put a default

File: trainers/server/src/trainers_server/dp_worker/main.py [hunk 2]
```diff
@@ -1,36 +1,165 @@
...
 
+        # Server has stopped — signal all workers to exit.
+        op_t = torch.tensor([255], dtype=torch.int32)  # OP_EXIT
+        if dist.is_available() and dist.is_initialized():
+            dist.broadcast(op_t, src=0)
+    else:
+        worker_loop(controller)
+
+    # ── Cleanup ───────────────────────────────────────────────────────
+    from megatron.core import parallel_state
```
Comment (rcano-baseten) [inline trainers/server/src/trainers_server/dp_worker/main.py:100]: is this inlline import necessary?

File: trainers/server/src/trainers_server/dp_worker/api/controller.py [hunk 3]
```diff
@@ -37,599 +39,671 @@
...
-
-            logger.info("forward_backward: done in %.2fs — loss=%.4f, num_samples=%d",
-                        time.perf_counter() - t0, self.last_loss, len(details.data))
-            return ForwardBackwardResult(
-                loss_fn_output_type="per_token_logprobs",
-                loss_fn_outputs=loss_fn_outputs,
-                metrics={"loss": self.last_loss},
-            )
+                raise ValueError(
+                    f"Unsupported loss_fn={details.loss_fn!r}. Only 'cross_entropy' is supported."
```
Comment (rcano-baseten) [inline trainers/server/src/trainers_server/dp_worker/api/controller.py:365]: probably worth doing a `", ".join(supported_loss_fns)` here so we don't have to keep this error message up to date. 

Also - ValueError is going to raise 500 - I'd suggest we make a ValidationError so that we can map to 400

File: trainers/server/src/trainers_server/dp_worker/api/controller.py [hunk 4]
```diff
@@ -37,599 +39,671 @@
...
-            return int(sock.getsockname()[1])
-
-    def _init_vllm_client(self) -> None:
-        deadline = time.time() + float(_ROLLOUT_STARTUP_TIMEOUT_SECONDS)
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            s.bind(("127.0.0.1", 0))
+            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            return int(s.getsockname()[1])
+
+    def _launch_rollout(self, model_path: str) -> None:
```
Comment (rcano-baseten) [inline trainers/server/src/trainers_server/dp_worker/api/controller.py:234]: do we need this anymore? Since we're doing our rollouts via samplers (not connected to the trainer), i'm thinking we won't need this anymore. But perhaps for a future PR

File: trainers/server/src/trainers_server/dp_worker/api/controller.py [hunk 5]
```diff
@@ -37,599 +39,671 @@
...
-            return float(value)
-        except (TypeError, ValueError) as exc:
-            raise ValueError("loss_fn_inputs.reward must be numeric.") from exc
+    # ── Batch construction ─────────────────────────────────────────────
+
+    def _build_batch(self, data: List[Datum]) -> Dict[str, torch.Tensor]:
+        """Build padded tensors from a list of Datum objects.
+
+        Returns a dict with keys:
+          input_ids   (B, S)     — input token IDs (padded)
```
Comment (rcano-baseten) [inline trainers/server/src/trainers_server/dp_worker/api/controller.py:305]: who's doing the padding?

File: trainers/server/src/trainers_server/dp_worker/api/controller.py [hunk 6]
```diff
@@ -37,599 +39,671 @@
...
+        """Generate samples from the vLLM rollout server.  Rank-0 only."""
         with self._lock:
             if self.mode != "inference":
-                raise RuntimeError("sample is only allowed in inference mode.")
+                raise RuntimeError("sample() is only valid in inference mode.")
+            if _rank() != 0:
+                return SampleResult(sequences=[])
+
             t0 = time.perf_counter()
+            self._ensure_tokenizer()
```
Comment (rcano-baseten) [inline trainers/server/src/trainers_server/dp_worker/api/controller.py:424]: we should be token-in token-out if I'm not mistaken

File: trainers/server/tests/test_controller_gpu.py [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+# ── Test: single GPU, forward_backward + optim_step, loss decreases ──────────
+
+
+def _single_gpu_rank(rank: int, world_size: int, dist_port: int, result_path: str) -> None:
+    """Single-GPU: verify forward_backward produces a finite loss and optim_step
+    reduces it over multiple iterations on the same batch."""
+    gpus = [0]
+    _init_dist(rank, world_size, dist_port, gpus[rank])
+
+    from trainers_server.dp_worker.api.controller import (
```
Comment (rcano-baseten) [inline trainers/server/tests/test_controller_gpu.py:94]: do we need the inline import? Is it preferred in this scenario for any reason?

File: trainers/server/tests/test_controller_gpu.py [hunk 8]
```diff
@@ -0,0 +1,559 @@
...
+                "update_successful": True,  # if we got here, it worked
+            }, f)
+    else:
+        worker_loop(controller)
+
+    _teardown()
+
+
+@skip_if_no_gpu(1)
+def test_single_gpu_forward_backward_and_optim():
```
Comment (rcano-baseten) [inline trainers/server/tests/test_controller_gpu.py:147]: for these tests, should be asserting that the loss is the same regardless of the parallelization straetegy? We should be able to make the assertions strict, as the calculations are largely deterministic (within some precision range)

File: trainers/server/tests/test_controller_gpu.py [hunk 9]
```diff
@@ -0,0 +1,559 @@
...
+        RLController, OP_EXIT, worker_loop
+    )
+    from trainers_server.dp_worker.api.models import (
+        RLControllerConfig, TrainingServerConfig, InferenceServerConfig
+    )
+
+    config = RLControllerConfig(
+        model_id=MODEL_PATH,
+        training=TrainingServerConfig(tensor_parallel_size=2, pipeline_parallel_size=1, gpus=gpus, max_length=64),
+        inference=InferenceServerConfig(tensor_parallel_size=1, gpus=[2]),
```
Comment (rcano-baseten) [inline trainers/server/tests/test_controller_gpu.py:342]: do we need inference in this test?

File: trainers/server/src/trainers_server/dp_worker/api/controller.py [hunk 10]
```diff
@@ -37,599 +39,671 @@
...
-            })
+
+            op_t = torch.tensor([OP_OPTIM_STEP], dtype=torch.int32)
+            _broadcast_tensor(op_t)
+
+            payload = pickle.dumps(details.model_dump())
+            _broadcast_bytes(payload)
+
+            result = self._execute_optim_step(details)
+            assert result is not None
```
Comment (rcano-baseten) [inline trainers/server/src/trainers_server/dp_worker/api/controller.py:392]: is this `assert result is not None` going to be easy to debug if we hit it? If this assert fails, what would be our next step?

**Review by rcano-baseten — COMMENTED** [review]
I'm a bit surprised we have to write code at this level. I figured we'd be able to instantiate a "trainer" that is owned by megatron / some framework, and then we'd be able execute some number of steps in a loop based on the micro batch size. 

Separately, I love the tests and the structure here. Let's chat about high level a bit (why this level of detail is necessary) and figure out what next steps are

File: trainers/server/src/trainers_server/dp_worker/api/controller.py [hunk 11]
```diff
@@ -37,599 +39,671 @@
...
+        # stage (post_process=True). Intermediate PP stages return hidden
+        # activations passed via pipeline communicators — we skip manual loss
+        # computation there. For PP=1, output is always (B, S).
+        if output_tensor is not None and output_tensor.shape == (B, S):
+            per_token = output_tensor.float() * loss_mask   # zero out padding
+            valid_counts = loss_mask.sum(dim=-1).clamp(min=1)   # (B,)
+            per_sample = per_token.sum(dim=-1) / valid_counts    # (B,)
+            loss = (per_sample * rewards).mean()
+        elif output_tensor is not None:
+            # Unexpected shape (e.g., intermediate PP activation) — fall back
```
Comment (rcano-baseten) [inline trainers/server/src/trainers_server/dp_worker/api/controller.py:539]: what shape would we expect in the case of intermediate PP activation? Is it impossible to predict? Can we guardrail this better?

File: trainers/server/src/trainers_server/dp_worker/api/controller.py [hunk 12]
```diff
@@ -37,599 +39,671 @@
...
+            # to a zero loss so backward is a no-op on this rank.
+            loss = torch.zeros(1, device=device, requires_grad=True)
+            loss = loss * output_tensor.sum() * 0  # keep graph connected
+        else:
```
Comment (rcano-baseten) [inline trainers/server/src/trainers_server/dp_worker/api/controller.py:543]: what does it mean for the output tensor to be None?

**Review by rcano-baseten — COMMENTED** [review]
just a couple thoughts here

File: trainers/server/src/trainers_server/dp_worker/api/controller.py [hunk 13]
```diff
@@ -37,599 +39,671 @@
...
-                use_hf=True,
+        self._closed: bool = False
+        self._rollout_process: Optional[subprocess.Popen] = None
+        self._tokenizer = None
+        self._pad_token_id: Optional[int] = None
+
+        # Find a rollout port on rank 0 and broadcast to all ranks.
+        # _broadcast_tensor handles NCCL's requirement for CUDA tensors.
+        port_t = torch.tensor([self._find_free_port() if _rank() == 0 else 0], dtype=torch.int64)
+        _broadcast_tensor(port_t)
```
Comment (rcano-baseten) [inline trainers/server/src/trainers_server/dp_worker/api/controller.py:164]: when we use `_broadcast_tensor` in this situation, what's happening? Why is it necessary?
